### PR TITLE
Wallet generate-address endpoint + refactors

### PR DESCRIPTION
**Pull request by anler**
# What

* Implement address generation in the wallet
* Added a `--testnet` flag to wallet's cli (and config) to tell the server it should generate /testnet/ addresses instead of /mainnet/ ones
* Refactor wallet architecture to merge all the synchronous actors into one: the Worker actors pool
* Refactor `witnet_crypto` and dependencies to use the `Protected` type for holding secrets
* Move business logic from the inside the Worker actor into independent structs inspired by the [Repository Pattern](https://www.youtube.com/watch?v=rtXpYpZdOzM)
* Isolate App actor state into its own structure in order to handle better its invariants
* Index UTXOs classified by account
* Calculate account balances
* Notify account balances after a new block is indexed

## Running the server

`$ cargo run -- --trace wallet server --testnet`

### Before

```

    +-------------+    +--------+
    | Controller  +--->|  App   |
    +-------------+    +--+-----+
                          |
          +---------------+---------------+
          |               |               |
          V               V               V
    +-----+-------+  +----+------+  +-----+-------+
    |  Storage    |  | Crypto    |  | RadExecutor |
    +-------------+  +-----------+  +-------------+
```

### After

Synchronous actors `Storage`, `Crypto` and `RadExecutor` has been merged into a synchronous `Worker` actor pool. This simplifies greatly the interaction between the different blocking operations (db access, network requests, and crypto operations)
```
    +-------------+    +--------+
    | Controller  +--->|  App   |
    +-------------+    +--+-----+
                          |
                          V
                     +----+------+
                     |  Worker   ||
                     +-----------+|
                      +-----------+
```

## FIXMEs

Currently, some db operations cannot be done atomically because the `rocksdb` crate doesn't support transactions yet (Rocksdb [supports them](https://github.com/facebook/rocksdb/wiki/Transactions) and [a PR](https://github.com/rust-rocksdb/rust-rocksdb/pull/250) is already opened) so we are using stdlib's Mutex in order to have exclusive access when updating some shared database fields

### CODE REVIEW

File: data_structures/src/chain.rs [hunk 1]
```diff
@@ -602,20 +615,37 @@ impl PublicKey {
...
 /// Secret Key data structure
-#[derive(Debug, Default, Eq, PartialEq, Clone, Serialize, Deserialize)]
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize)]
 pub struct SecretKey {
-    // FIXME(#560): Use Protected type
-    pub bytes: [u8; 32],
+    pub bytes: Protected,
+}
+
+// FIXME: SecretKey shouldn't implement Default
```
Comment (tmpolaczyk) [inline data_structures/src/chain.rs]: Why not?

File: wallet/src/account.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/// Result type for accounts-related operations that can fail.
+pub type Result<T> = std::result::Result<T, failure::Error>;
+
+/// Root KeyPath used for wallet accounts.
+#[inline]
+pub fn account_keypath(index: u32) -> types::KeyPath {
+    types::KeyPath::default()
+        .hardened(3)
+        .hardened(4919)
+        .hardened(index)
```
Comment (tmpolaczyk) [inline wallet/src/account.rs]: Why 3 and 4919?

Comment (anler) [reply wallet/src/account.rs]: It was defined here: https://github.com/witnet/sheikah/blob/master/app/main/api/handlers/generateAddress.ts#L110

File: config/src/config.rs [hunk 3]
```diff
@@ -494,16 +502,21 @@ pub struct Wallet {
...
+    /// Duration in milliseconds after which outgoing request should timeout.
+    pub requests_timeout: u64,
 }
 
 impl Wallet {
     pub fn from_partial(config: &PartialWallet, defaults: &dyn Defaults) -> Self {
         Wallet {
+            testnet: config.testnet,
             session_expires_in: config.session_expires_in.unwrap_or_else(|| 3200),
+            requests_timeout: config.requests_timeout.unwrap_or_else(|| 60_000),
```
Comment (mariocao) [inline config/src/config.rs]:
```suggestion
            requests_timeout: config.requests_timeout.unwrap_or(60_000),
```

File: config/src/config.rs [hunk 4]
```diff
@@ -494,16 +502,21 @@ pub struct Wallet {
...
     pub session_expires_in: u64,
+    /// Duration in milliseconds after which outgoing request should timeout.
+    pub requests_timeout: u64,
 }
 
 impl Wallet {
     pub fn from_partial(config: &PartialWallet, defaults: &dyn Defaults) -> Self {
         Wallet {
+            testnet: config.testnet,
             session_expires_in: config.session_expires_in.unwrap_or_else(|| 3200),
```
Comment (mariocao) [inline config/src/config.rs]:
```suggestion
            session_expires_in: config.session_expires_in.unwrap_or(3200),
```

File: wallet/src/account.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
+use crate::types;
+
+/// Result type for accounts-related operations that can fail.
+pub type Result<T> = std::result::Result<T, failure::Error>;
+
+/// Root KeyPath used for wallet accounts.
+#[inline]
+pub fn account_keypath(index: u32) -> types::KeyPath {
```
Comment (mariocao) [inline wallet/src/account.rs]: Perhaps we should add a link to the WIP: https://github.com/aesedepece/WIPs/blob/wip-adansdpc-hdwallets/wip-adansdpc-hdwallets.md

Comment (anler) [reply wallet/src/account.rs]: I'll add this https://github.com/aesedepece/WIPs/blob/wip-adansdpc-hdwallets/wip-adansdpc-hdwallets.md#path-levels

File: wallet/src/account.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        let keypath = types::KeyPath::default().index(0);
+
+        account_key.derive(engine, &keypath)?
+    };
+    let internal = {
+        let keypath = types::KeyPath::default().index(1);
+
+        account_key.derive(engine, &keypath)?
+    };
+    let rad = {
```
Comment (mariocao) [inline wallet/src/account.rs]: You may remove the rad path as it is not used! :)

File: wallet/src/constants.rs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+/// Default offset used when returning paginated results.
+pub static DEFAULT_OFFSET: u32 = 0;
```
Comment (mariocao) [inline wallet/src/constants.rs]:
```suggestion
pub static DEFAULT_PAGINATION_OFFSET: u32 = 0;
```

File: wallet/src/constants.rs [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+
+/// Default limit/page-size used when returning paginated results.
+pub static DEFAULT_LIMIT: u32 = 25;
```
Comment (mariocao) [inline wallet/src/constants.rs]:
```suggestion
pub static DEFAULT_PAGINATION_LIMIT: u32 = 25;
```

File: wallet/src/constants.rs [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+
+/// Maximum limit/page-size that can be used when returning paginated
+/// results.
+pub static MAX_LIMIT: u32 = 1000;
```
Comment (mariocao) [inline wallet/src/constants.rs]:
```suggestion
pub static MAX_PAGINATION_LIMIT: u32 = 1000;
```

File: wallet/src/actors/worker/methods.rs [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        SyncArbiter::start(concurrency, move || Self {
+            db: db.clone(),
+            wallets: wallets.clone(),
+            params: params.clone(),
+            rng: rand_os::OsRng,
+            engine: engine.clone(),
+        })
+    }
+
+    pub fn run_rad_request(&self, request: types::RADRequest) -> Result<types::RadonTypes> {
```
Comment (tmpolaczyk) [inline wallet/src/actors/worker/methods.rs]: I think all the data request logic should not be implemented here, the witnet_radon crate should expose a `run_rad_request` method which would return a `Future<Item=RadonTypes, Error=RadonError>`.

File: wallet/src/actors/worker/methods.rs [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            &self.params.id_hash_function,
+            &master_key,
+            self.params.master_key_salt.as_ref(),
+            self.params.id_hash_iterations,
+        );
+
+        let default_account_index = 0;
+        let default_account =
+            account::gen_account(&self.engine, default_account_index, &master_key)?;
+
```
Comment (tmpolaczyk) [inline wallet/src/actors/worker/methods.rs]: Add comment:

```suggestion
// This is for storage encryption
```

File: wallet/src/actors/worker/methods.rs [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        let prefix = id.as_bytes().to_vec();
+        let salt = crypto::salt(&mut self.rng, self.params.db_salt_length);
+        let iv = crypto::salt(&mut self.rng, self.params.db_iv_length);
+        let key = crypto::key_from_password(password, &salt, self.params.db_hash_iterations);
+
+        let wallet_db = db::EncryptedDb::new(self.db.clone(), prefix, key, iv.clone());
+        wallet_db.put("", ())?; // used when unlocking to check if the password is correct
```
Comment (tmpolaczyk) [inline wallet/src/actors/worker/methods.rs]:
```suggestion
        wallet_db.put(ENCRYPTION_CHECK_KEY, ENCRYPTION_CHECK_MAGIC)?; // used when unlocking to check if the password is correct
```

File: wallet/src/actors/worker/methods.rs [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        let session_id = crypto::gen_session_id(
+            &mut self.rng,
+            &self.params.id_hash_function,
+            &key,
+            &salt,
+            self.params.id_hash_iterations,
+        );
+        let prefix = wallet_id.as_bytes().to_vec();
+        let wallet_db = db::EncryptedDb::new(self.db.clone(), prefix, key, iv);
+        wallet_db.get::<_, ()>("").map_err(|err| match err {
```
Comment (tmpolaczyk) [inline wallet/src/actors/worker/methods.rs]:
```suggestion
        wallet_db.get::<_, String>(ENCRYPTION_CHECK_KEY).map_err(|err| match err {
```

And verify that the result matches `ENCRYPTION_CHECK_MAGIC`

**Comment by mariocao** [comment]
I would include some instructions in the `README.md` for those that are lazy enough not to look into the repo's wiki.

**Comment by mariocao** [comment]
If possible, I would not log session IDs. Their use is similar to access tokens.

File: wallet/src/actors/worker/methods.rs (hunk 10 above)
Comment (anler) [reply wallet/src/actors/worker/methods.rs]: I agree, although the implementation @aesedepece and @Tommytrg had done is not async so it doesn't work with futures, maybe we could open an issue to address later, what do you think @mariocao ?

**Comment by anler** [comment]
@mariocao can you re-review please?

**Pull request merged**